zookeeper/src/main/java/com/linecorp/armeria/server/zookeeper/ZooKeeperUpdatingListener.java

```diff
@@ -62,57 +133,122 @@ public static ZooKeeperUpdatingListenerBuilder builder(CuratorFramework client,
+        if (spec instanceof EndpointInstanceSpec) {
+            final Endpoint endpoint = ((EndpointInstanceSpec) spec).endpoint();
+            if (endpoint.hasPort() && validatePort(server, endpoint.port(), null)) {
+                return spec;
+            }
+            final ServerPort serverPort = server.activePort();
+            assert serverPort != null;
+            return InstanceSpec.ofEndpoint(endpoint.withPort(serverPort.localAddress().getPort()));
+        } else if (spec instanceof CuratorXInstanceSpec) {
+            final ServiceInstance<?> serviceInstance = ((CuratorXInstanceSpec) spec).serviceInstance();
+            return fillAndCreateNewInstanceSpec(serviceInstance, server);
+        } else {
+            return spec;
+        }
+    }
+
+    private static InstanceSpec fillAndCreateNewInstanceSpec(
+            ServiceInstance<?> serviceInstance, Server server) {
+        final CuratorXInstanceSpecBuilder builder =
+                InstanceSpec.curatorXInstanceBuilder(serviceInstance.getName());
+        builder.serviceId(serviceInstance.getId());
+        final String address;
+        if (serviceInstance.getAddress() != null) {
+            address = serviceInstance.getAddress();
+        } else {
+            final Inet4Address inet4Address = SystemInfo.defaultNonLoopbackIpV4Address();
+            address = inet4Address != null ? inet4Address.getHostAddress() : server.defaultHostname();
+        }
+        builder.serviceAddress(address);
+        final int port = port(server, SessionProtocol.HTTP, serviceInstance.getPort());
+        if (port > 0) {
+            builder.port(port);
+        }
+        final int sslPort = port(server, SessionProtocol.HTTPS, serviceInstance.getSslPort());
+        if (sslPort > 0) {
+            builder.sslPort(sslPort);
+        }
+        builder.serviceType(serviceInstance.getServiceType());
+        final Object payload = serviceInstance.getPayload();
+        if (payload != null) {
+            builder.payload(payload);
+        }
+        return builder.build();
+    }
+
+    private static int port(Server server, SessionProtocol protocol, @Nullable Integer port) {
+        if (port != null) {
+            if (validatePort(server, port, protocol)) {
+                return port;
+            }
+        }
+        final ServerPort serverPort = server.activePort(protocol);
+        return serverPort != null ? serverPort.localAddress().getPort() : -1;
+    }
+
+    private static boolean validatePort(Server server, int port, @Nullable SessionProtocol protocol) {
+        for (ServerPort serverPort : server.activePorts().values()) {
+            if ((protocol == null || serverPort.hasProtocol(protocol)) &&
+                serverPort.localAddress().getPort() == port) {
+                return true;
+            }
+        }
+        logger.warn("The port number: {} (expected one of activePorts: {})",
+                    port, server.activePorts());
```

Comment:
Should we log this or not?

Reply:
The user specified a certain port but the server does not open that port.
So I think I should warn.
Perhaps, it's better to rephrase the sentence?

Reply:
Rephrasing SGTM